src/net.cpp

```diff
@@ -719,6 +723,173 @@ CNetMessage V1TransportDeserializer::GetMessage(const CMessageHeader::MessageSta
+        const unsigned int remaining = m_message_size + CHACHA20_POLY1305_AEAD_TAG_LEN - m_data_pos;
+        const unsigned int copy_bytes = std::min(remaining, bytes);
+
+        // extend buffer, respect previous copied AAD part
+        if (vRecv.size() < CHACHA20_POLY1305_AEAD_AAD_LEN + m_data_pos + copy_bytes) {
+            // Allocate up to 256 KiB ahead, but never more than the total message size (incl. AAD & TAG).
+            vRecv.resize(std::min(CHACHA20_POLY1305_AEAD_AAD_LEN + m_message_size + CHACHA20_POLY1305_AEAD_TAG_LEN, CHACHA20_POLY1305_AEAD_AAD_LEN + m_data_pos + copy_bytes + 256 * 1024 + CHACHA20_POLY1305_AEAD_TAG_LEN));
+        }
+
+        memcpy(&vRecv[CHACHA20_POLY1305_AEAD_AAD_LEN + m_data_pos], pch, copy_bytes);
+        m_data_pos += copy_bytes;
+
+        return copy_bytes;
+    }
+}
+
+CNetMessage V2TransportDeserializer::GetMessage(const CMessageHeader::MessageStartChars& message_start, int64_t time)
+{
+    // In v2, vRecv contains the encrypted payload plus the MAC tag (1+bytes serialized message command + ? bytes message payload + 16 byte mac tag)
+    assert(Complete());
+
+    // defensive decoding (MAC check, decryption, command deserialization)
+    // we'll always return a CNetMessage (even if encryption fails), we always increase the AEAD sequence numbers
+    bool valid_checksum = false;
+    bool valid_header = false;
+    std::string command_name;
+
+    // count bytes we decrypted including MAC tag + AD
+    m_bytes_decrypted += vRecv.size();
+
+    if (m_aead->Crypt(m_payload_seqnr, m_aad_seqnr, m_aad_pos, (unsigned char*)vRecv.data(), vRecv.size(), (const uint8_t*)vRecv.data(), vRecv.size(), false)) {
+        // MAC check was successful
+        valid_checksum = true;
+
+        // okay, we could decrypt it, now remove packet length and MAC tag
+        assert(vRecv.size() > CHACHA20_POLY1305_AEAD_AAD_LEN + CHACHA20_POLY1305_AEAD_TAG_LEN);
+        // CDataStream::erase at the begin will just increase the read pos
+        vRecv.erase(vRecv.begin(), vRecv.begin() + CHACHA20_POLY1305_AEAD_AAD_LEN);
+        vRecv.erase(vRecv.end() - CHACHA20_POLY1305_AEAD_TAG_LEN, vRecv.end());
+
+        uint8_t size_or_shortid = 0;
+        try {
+            vRecv >> size_or_shortid;
+        } catch (const std::ios_base::failure&) {
+            LogPrint(BCLog::NET, "Invalid command name\n");
+        }
+        if (size_or_shortid > 0 && size_or_shortid <= NET_P2P_V2_CMD_MAX_CHARS_SIZE && vRecv.size() >= size_or_shortid) {
+            // first byte is a number between 1 and 12. Must be a string command.
+            valid_header = true;
+
+            // use direct read since we already read the varlens size uint8_t
+            command_name.resize(size_or_shortid);
+            vRecv.read(&command_name[0], size_or_shortid);
+        }
+        // try for short ID in case the first byte is a number larger than 12
+        if (!valid_header && size_or_shortid >= 0) {
```

Comment:
`size_or_shortid` is `uint8_t`. Since it is unsigned, `size_or_shortid >= 0` will always be true. Should this be `if (!valid_header && size_or_shortid > 0) {`?

Reply:
Not sure. If `size_or_shortid` is `0` (due to a `std::ios_base::failure` or if `vRecv` contain `0`), I'd like to enter this `if` to set `command_name` to `unknown-0`.
Otherwise `command_name` would be unset for a size of `0` (which is somehow a undefined short-id).

Reply:
Ok, if I understand correctly, maybe:
```diff
-    bool valid_header = false;
     std::string command_name;
@@ -812,15 +811,12 @@ CNetMessage V2TransportDeserializer::GetMessage(const CMessageHeader::MessageSta
         }
         if (size_or_shortid > 0 && size_or_shortid <= NET_P2P_V2_CMD_MAX_CHARS_SIZE && vRecv.size() >= size_or_shortid) {
             // first byte is a number between 1 and 12. Must be a string command.
-            valid_header = true;
 
             // use direct read since we already read the varlens size uint8_t
             command_name.resize(size_or_shortid);
             vRecv.read(&command_name[0], size_or_shortid);
-        }
-        // try for short ID in case the first byte is a number larger than 12
-        if (!valid_header && size_or_shortid >= 0) {
-            valid_header = true;
+        } else if (size_or_shortid == 0 || size_or_shortid > NET_P2P_V2_CMD_MAX_CHARS_SIZE) {
+            // try for short ID if first byte is 0 or a number larger than 12
             if (!GetCommandFromShortCommandID(size_or_shortid, command_name)) {
@@ -842,7 +838,7 @@ CNetMessage V2TransportDeserializer::GetMessage(const CMessageHeader::MessageSta
     msg.m_command = command_name;
 
     // store state about valid header, netmagic and checksum
-    msg.m_valid_header = valid_header; // not relevant for v2, always pass
+    msg.m_valid_header = true ; // not relevant for v2, always pass
```

Reply:
or perhaps even just `else`
```diff
-    bool valid_header = false;
     std::string command_name;
 
     // count bytes we decrypted including MAC tag + AD
@@ -812,15 +811,12 @@ CNetMessage V2TransportDeserializer::GetMessage(const CMessageHeader::MessageSta
         }
         if (size_or_shortid > 0 && size_or_shortid <= NET_P2P_V2_CMD_MAX_CHARS_SIZE && vRecv.size() >= size_or_shortid) {
             // first byte is a number between 1 and 12. Must be a string command.
-            valid_header = true;
 
             // use direct read since we already read the varlens size uint8_t
             command_name.resize(size_or_shortid);
             vRecv.read(&command_name[0], size_or_shortid);
-        }
-        // try for short ID in case the first byte is a number larger than 12
-        if (!valid_header && size_or_shortid >= 0) {
-            valid_header = true;
+        } else {
+            // try for short ID if first byte is 0 or a number larger than NET_P2P_V2_CMD_MAX_CHARS_SIZE
             if (!GetCommandFromShortCommandID(size_or_shortid, command_name)) {
                 // unknown-short-id
                 //  results in a valid but unknown message (will be skipped)
@@ -842,7 +838,7 @@ CNetMessage V2TransportDeserializer::GetMessage(const CMessageHeader::MessageSta
     msg.m_command = command_name;
 
     // store state about valid header, netmagic and checksum
-    msg.m_valid_header = valid_header; // not relevant for v2, always pass
+    msg.m_valid_header = true ; // not relevant for v2, always pass
```

Reply:
What if the buffer contains a `0x09` as first bytes (a valid string based command with the size of 9 chars) but those 9 bytes would actually not follow? Wouldn't it become `valid_header=true` in your suggestions?

Reply:
You're right; I deleted my second suggestion. Unless I'm confused I think the first suggestion was correct for the case you describe but can be better:
```cpp
if (vRecv.size() >= size_or_shortid) {
    // Buffer length is at least as large as the first byte defining the length of the command string
    if (size_or_shortid > 0 && size_or_shortid <= NET_P2P_V2_CMD_MAX_CHARS_SIZE) {
        // First byte is a number between 1 and 12. Must be a string command.
        // Use direct read since we already read the varlens size uint8_t
        command_name.resize(size_or_shortid);
        vRecv.read(&command_name[0], size_or_shortid);
    } else {
        // Not a string command; try for short ID
        if (!GetCommandFromShortCommandID(size_or_shortid, command_name)) {
            // unknown-short-id results in a valid but unknown message (will be skipped)
            command_name = "unknown-" + ToString(size_or_shortid);
        }
    }
}
```

I'm unsure what purpose `valid_header` served, as it would always become true at the latest by line 823. Is a false value relevant for `m_valid_header` at line 845, or can it be set to true?
```
msg.m_valid_header = valid_header; // not relevant for v2, always pass
```

Unrelated, but it looks like short command IDs need to be added for the `GETCFCHECKPT` and `CFCHECKPT` protocol message types that were added in f9e00bb25ac.

Reply:
You're right. I deleted the second suggestion. Unless I'm confused, I think the suggestion above is still correct for the case you describe; `net_tests` and the `p2p_v2_transport_deserializer` fuzzer both pass.

It's not clear to me what purpose `valid_header` serves because it always becomes true at the latest by line 823.

Unrelated, but it looks like short command IDs need to be added for the `GETCFCHECKPT` and `CFCHECKPT` BIP157 message types that were added in f9e00bb25ac and possibly soon for `GETCFHEADERS` and `GETCFFILTERS` as well.

Reply:
Agreed this is a bit confusing, and I'm not exactly sure what you're trying to achieve. `size_or_shortid >= 0` will *always* evaluate to true. `valid_header` will *always* eventually be set to true. Under what circumstances do you think `valid_header` should be false?

Reply:
I just removed the `size_or_shortid >= 0` check (your right,.. it's always true).
The `(!valid_header)` check is necessary though,... in case `vRecv >> size_or_shortid` failed (can't happen?!) or if it contains a value greater than 12 or if below 12, the remaining buffer doesn't have the expected size.